src/main/java/picocli/CommandLine.java

```diff
@@ -11059,7 +11081,8 @@ static CommandSpec extractCommandSpec(Object command, IFactory factory, boolean
 
                 result.updateArgSpecMessages();
 
-                if (annotationsAreMandatory) {validateCommandSpec(result, hasCommandAnnotation, userObject.toString()); }
+                if (annotationsAreMandatory) {validateCommandSpec(result, cmd != null || hasCommandAnnotation, userObject.toString()); }
+                result = DefaultFactory.create(defaultFactory(), cmd == null  ? NoOpPreprocessor.class :  cmd.preprocessor()).preprocess(result);
```

Comment:
Just thinking out loud: perhaps we should first construct the full command hierarchy, and then have a second pass over the hierarchy (recursively visiting subcommands and sub-subcommands) to apply the preprocessor?

That way the preprocessor can operate on the full hierarchy and there are no timing issues where the preprocessor is run before the hierarchy has been fully constructed.

Reply:
Well I think that's not really necessery, where do you see this more fit ? What kind of timing issues ? I'm open to suggestions, if there is any appearent improvement :)

Reply:
Oops, I only just saw your comment now.
What I meant was, the current proposal applies the transformation immediately after the command was created, so while the construction of the full hierarchy is still in progress. Imagine we have this hierarchy:

```
Top
  - Sub1
    - Sub1Sub1
    - Sub1Sub2
  - Sub2
    -Sub2Sub1
    -Sub2Sub2
```

The transformation for `Sub1Sub1` will be applied before `Sub1Sub2` was created, and even before `Sub2` and its subcommands are created. So if the transformation for  `Sub1Sub1` wanted to do something with  `Sub1Sub2`, that is currently not possible.

The current implementation does not allow some transformations, because the command hierarchy has not been fully constructed yet.

 I think it would be better if transformations were applied after the command hierarchy has been completely constructed.

Reply:
@remkop I don't think it's a good practice. This is not localized . And all preprocessing/transformation depends only on external state, so if there are common things, they could be unified by parent transformer class. The transformer doesn't have to be local to the subcommand.

Reply:
@MarkoMackic well, we don't want to fail some people's transformations because they do things differently than we think they should... I think the transformation should not be applied to a partial model. Transformations should be able to trust that the model is complete when the `transform` method is called.

Reply:
@remkop ok, let's try it, f.y.i, it shouldn't affect current API ( it should be still context aware ), but I can make it traverse the model after it's built. But there are issues with that, sp let's first communicate those. Let's take your tree for example :
```
Top
  - Sub1
    - Sub1Sub1
    - Sub1Sub2
  - Sub2
    -Sub2Sub1
    -Sub2Sub2
```

So let's say we define 2 transformers 1 for Sub1 and second for Sub2,  Sub1 transformer moves subcommand Sub2Sub1 ( eg. Sub1 transformer would look something like this ( pseudo code ) `spec.getParentCommand().subcommands("Sub2").subcommands("Sub2Sub1").moveTo(spec)` , and than you have transformer in Sub2 that says `spec.removeSubcommand("Sub2Sub1")`. They are clashing, 2 ways possible, if Sub1 transformer is executed first, the second one will fail, , and the other way around  , and it's up to library user to do that sensitive handling. Operating on partial model is much more stable, your model preprocessor can rest assured that hierarchy down the tree from what you're processing is 100% built, can't operate on parallel or branches up the tree. I think this is just more secure approach, and there are no API limitations, just some rules are enforced, ( parent can only process it's children and itself )

Reply:
I see a couple of advantages over transforming done model, but they are not in out-of-scope modification, they could be just to inspect the model and make decisions. But even that is relatively hard, let's imagine with the same tree 
```
Top
  - Sub1
    - Sub1Sub1
    - Sub1Sub2
  - Sub2
    -Sub2Sub1
    -Sub2Sub2
```


So again we have model transformer on Sub1 and Sub2, and model transformer at Sub1 says something like this

`(spec) -> if (spec.getParentCommand().subcommands("Sub2").hasSubcommand("Sub2Sub1") spec.removeSubcommand("Sub1Sub2") ` 

and model transformer on Sub2 says 

`(spec) -> if (Boolean.valueOf(System.getProp("whatever"))) spec.removeSubcommand("Sub2Sub1")`

but than we'd have to construct dependency tree, because Sub2 transformer has to run before Sub1 transformer etc.. and it wouldn't work for cyclic deps .. much work for in my opinion small benefit.

Reply:
If we were to implement transformation after the model is completely constructed, how would we go about it?

Reply:
@remkop we'll recursive tree traversal inside constructor, after spec creation ? maybe provide a method to replace spec if there isn't one?

Reply:
But what do you think about the notes above ? is it worth it ? We shouldn't rush.

Reply:
I think your argument is basically that, even if we do transformations after the model is fully built, that does not eliminate the possibility for people to get themselves into trouble (for example by having 2 transformations that operate on the same node).

You may be right.
My intuition is that the "right" thing to do is to wait with transformations until the model is fully built.
One advantage is that it makes it possible to (in the future) have different traversal orders: breadth-first, depth-first, and variations of these. 
It just does not feel right to do transformations while the model is not fully built, sorry if that is vague.

Is the effort that much? I think the key question is where to put the call to start the transformations.

Reply:
Ok, I'll refactor due to your request, but I hold my attitude that it's a wrong decision :)